### ISSUE TITLE
GEMM GPU performance: slower than CPU at all matrix sizes

### ISSUE THREAD

**Issue by grillcheese123** [Grillcheese-AI/grilly#7]
## Problem

GPU GEMM is significantly slower than CPU for all tested matrix sizes on AMD RX 6750 XT:

```
  Size    GPU ms    GFLOP/s    CPU ms   Speedup
----------------------------------------------------
   128     0.50       8.4     0.07     0.14x
   256     1.00      33.7     0.25     0.26x
   512     2.23     120.6     0.76     0.34x
  1024    14.39     149.3     4.71     0.33x
  2048    69.12     248.6    40.39     0.58x
```

At 2048x2048, GPU achieves only 248 GFLOP/s — the RX 6750 XT should theoretically deliver ~13 TFLOP/s FP32. We're at <2% utilization.

### 1. Buffer allocation overhead
Every GEMM call allocates new Vulkan buffers via `_acquire_buffer` / `_release_buffer`. This incurs:
- `vkAllocateMemory` + `vkBindBufferMemory` per buffer (expensive)
- No buffer reuse between calls
- No persistent memory pool

**Fix**: Implement a VMA-style buffer pool or use [pyvma](https://github.com/realitix/pyvma) for persistent buffer allocation. Pre-allocate input/output buffers once and reuse across GEMM calls.

### 2. Per-call pipeline/descriptor overhead
`get_or_create_pipeline` and `get_cached_descriptor_set` are called per GEMM. Even with caching, descriptor set allocation and command buffer submission have fixed overhead.

**Fix**: Batch multiple GEMM calls into a single command buffer. Use descriptor set pools with pre-allocated sets.

### 3. Upload/download synchronization
Each GEMM call does:
1. Upload A to GPU (`_upload_buffer`)
2. Upload B to GPU (`_upload_buffer`)
3. Submit compute
4. Wait for completion
5. Download C from GPU (`_download_buffer`)

Steps 1-2 and 5 are synchronous PCIe transfers with full fence waits.

**Fix**: 
- Use staging buffers with async transfers
- Keep tensors on GPU between operations (VulkanTensor)
- Use double-buffered upload/compute overlap

### 4. Shader efficiency
Current `gemm_tiled.glsl` uses 4x4 register blocking with workgroup size 16x16. For AMD RDNA2:
- Shared memory tiling may need tuning for 64-wide wavefronts
- Consider 8x8 register blocking
- Investigate subgroup operations for intra-wavefront reduction

### 5. Missing features
- No batched GEMM (single dispatch for multiple small matmuls)
- No persistent kernel mode for repeated same-size GEMMs
- No support for transposed inputs (currently requires explicit `np.ascontiguousarray(W.T)`)

## Impact

This blocks GPU-accelerated VSA head training. The cached training loop calls `fused_linear_tanh` (GEMM + activation) and GEMM for weight gradients — both would be faster on CPU at current sizes (B=4-64, d_model=768, vsa_dim=10000).

## Proposed Plan

1. **Phase 1**: Implement buffer pooling (reuse allocations, pre-sized pools for common shapes)
2. **Phase 2**: Persistent VulkanTensor mode — keep weights on GPU, only upload activations
3. **Phase 3**: Async staging buffers with compute/transfer overlap
4. **Phase 4**: Shader tuning (larger register tiles, RDNA2-specific optimizations)
5. **Phase 5**: Batched GEMM and command buffer batching

## Environment
- GPU: AMD Radeon RX 6750 XT (RDNA2)
- Vulkan 1.3
- OS: Windows 10
- Benchmark: `examples/benchmark_gemm.py`
